site/app/controllers/ChatroomController.php

```diff
@@ -0,0 +1,234 @@
+            return JsonResponse::getFailResponse('Invalid Chatroom ID');
+        }
+
+        foreach ($chatroom->getMessages() as $message) {
+            $em->remove($message);
+        }
+        $em->remove($chatroom);
+        $em->flush();
+
+        return JsonResponse::getSuccessResponse();
+    }
+
+    /**
+     * @Route("/courses/{_semester}/{_course}/chat/{chatroom_id}/edit", methods={"POST"})
+     * @AccessControl(role="INSTRUCTOR")
+     */
+    public function editChatroom(int $chatroom_id): RedirectResponse {
+        $em = $this->core->getCourseEntityManager();
+        $chatroom = $em->getRepository(Chatroom::class)->find($chatroom_id);
+
+        if (!$chatroom) {
+            $this->core->addErrorMessage("Chatroom not found");
+            return new RedirectResponse($this->core->buildCourseUrl(['chat']));
+        }
+
+        if (isset($_POST['title'])) {
+            $chatroom->setTitle($_POST['title']);
+        }
+        if (isset($_POST['description'])) {
+            $chatroom->setDescription($_POST['description']);
+        }
+        if (isset($_POST['allow-anon'])) {
+            $chatroom->setAllowAnon(true);
+        }
+        else {
+            $chatroom->setAllowAnon(false);
+        }
+
+        $em->flush();
+
+        $this->core->addSuccessMessage("Chatroom successfully updated");
+        return new RedirectResponse($this->core->buildCourseUrl(['chat']));
+    }
+
+    /**
+     * @Route("/courses/{_semester}/{_course}/chat/{chatroom_id}/toggleOnOff", methods={"POST"})
+     * @AccessControl(role="INSTRUCTOR")
+     */
+     public function toggleChatroomOnOf(int $chatroom_id): RedirectResponse {
+        $em = $this->core->getCourseEntityManager();
+        $chatroom = $em->getRepository(Chatroom::class)->find($chatroom_id);
+
+        if ($chatroom === null) {
+            $this->core->addErrorMessage("Chatroom not found");
+            return new RedirectResponse($this->core->buildCourseUrl(['chat']));
+        }
+
+        $chatroom->toggle_on_off();
+
+        $em->flush();
+
+        return new RedirectResponse($this->core->buildCourseUrl(['chat']));
+     }
+
+
+    /**
+     * @Route("/courses/{_semester}/{_course}/chat/{chatroom_id}/messages", methods={"GET"})
+     */
+    public function fetchMessages(int $chatroom_id): JsonResponse {
+        $em = $this->core->getCourseEntityManager();
+        $messages = $em->getRepository(Message::class)->findBy(['chatroom' => $chatroom_id], ['timestamp' => 'ASC']);
+
+        $formattedMessages = array_map(function ($message) {
+            return [
+                'id' => $message->getId(),
+                'content' => $message->getContent(),
+                'timestamp' => $message->getTimestamp()->format('Y-m-d H:i:s'),
+                'user_id' => $message->getUserId(),
+                'display_name' => $message->getDisplayName(),
+                'role' => $message->getRole()
+            ];
+        }, $messages);
```

Comment:
I believe using `toArray()` on the Doctrine collection will have the same effect.